File: backend/src/models/MedicalCertificate.js
const { DataTypes } = require('sequelize');

module.exports = (sequelize) => {
  const MedicalCertificate = sequelize.define('MedicalCertificate', {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true
    },
    userId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'Users',
        key: 'id'
      }
    },
    // Datos del certificado
    certificateNumber: {
      type: DataTypes.STRING,
      allowNull: true
    },
    issueDate: {
      type: DataTypes.DATE,
      allowNull: false
    },
    startDate: {
      type: DataTypes.DATE,
      allowNull: false
    },
    endDate: {
      type: DataTypes.DATE,
      allowNull: false
    },
    requestedDays: {
      type: DataTypes.INTEGER,
      allowNull: false
    },
    // Información médica
    diagnosisCode: {
      type: DataTypes.STRING,
      allowNull: true
    },
    diagnosis: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    symptoms: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    // Centro médico
    hasVisitedDoctor: {
      type: DataTypes.BOOLEAN,
      allowNull: false,
      defaultValue: false
    },
    medicalCenter: {
      type: DataTypes.STRING,
      allowNull: true
    },
    attendingPhysician: {
      type: DataTypes.STRING,
      allowNull: true
    },
    medicalPrescription: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    // Cuestionario médico (JSON)
    questionnaire: {
      type: DataTypes.JSON,
      allowNull: true
    },
    // Estado y respuestas
    status: {
      type: DataTypes.ENUM('pending', 'under_review', 'approved', 'rejected', 'needs_audit'),
      defaultValue: 'pending'
    },
    // Respuesta del médico auditor
    auditorId: {
      type: DataTypes.UUID,
      allowNull: true,
      references: {
        model: 'Users',
        key: 'id'
      }
    },
    auditorResponse: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    
    // Información del médico tratante
    treatingPhysician: {
      type: DataTypes.STRING,
      allowNull: true,
      comment: 'Nombre del médico que atendió al empleado'
    },
    
    treatingPhysicianLicense: {
      type: DataTypes.STRING,
      allowNull: true,
      comment: 'Matrícula profesional del médico tratante'
    },
    
    medicalInstitution: {
      type: DataTypes.STRING,
      allowNull: true,
      comment: 'Institución médica donde fue atendido'
    },
    
    // Notificación a ART
    notifyART: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      comment: 'Si debe notificarse a la ART'
    },
    
    artNotified: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      comment: 'Si ya se notificó a la ART'
    },
    
    artNotificationDate: {
      type: DataTypes.DATE,
      allowNull: true,
      comment: 'Fecha de notificación a la ART'
    },
    approvedDays: {
      type: DataTypes.INTEGER,
      allowNull: true
    },
    needsAudit: {
      type: DataTypes.BOOLEAN,
      defaultValue: true
    },
    isJustified: {
      type: DataTypes.BOOLEAN,
      allowNull: true
    },
    auditDate: {
      type: DataTypes.DATE,
      allowNull: true
    },
    // Archivos adjuntos
    attachments: {
      type: DataTypes.JSON,
      allowNull: true,
      comment: 'URLs de archivos adjuntos como certificados escaneados'
    },
    // Metadata
    createdBy: {
      type: DataTypes.UUID,
      allowNull: false
    },
    lastModifiedBy: {
      type: DataTypes.UUID,
      allowNull: true
    }
  }, {
    tableName: 'medical_certificates',
    timestamps: true,
    indexes: [
      {
        fields: ['userId']
      },
      {
        fields: ['status']
      },
      {
        fields: ['auditorId']
      },
      {
        fields: ['issueDate']
      }
    ]
  });

  // Asociaciones
  MedicalCertificate.associate = (models) => {
    // Un certificado pertenece a un usuario
    MedicalCertificate.belongsTo(models.User, {
      foreignKey: 'userId',
      as: 'employee'
    });

    // Un certificado puede tener un médico auditor
    MedicalCertificate.belongsTo(models.User, {
      foreignKey: 'auditorId',
      as: 'auditor'
    });

    // Un certificado puede tener múltiples recetas
    MedicalCertificate.hasMany(models.MedicalPrescription, {
      foreignKey: 'certificateId',
      as: 'prescriptions'
    });
  };

  return MedicalCertificate;
};

File: frontend_flutter/lib/models/medical_certificate.dart
class MedicalCertificate {
  final String id;
  final String userId;
  final String? certificateNumber;
  final DateTime issueDate;
  final DateTime startDate;
  final DateTime endDate;
  final int requestedDays;
  
  // Información médica
  final String? diagnosisCode;
  final String? diagnosis;
  final String? symptoms;
  
  // Centro médico
  final bool hasVisitedDoctor;
  final String? medicalCenter;
  final String? attendingPhysician;
  final String? medicalPrescription;
  
  // Cuestionario médico
  final Map<String, dynamic>? questionnaire;
  
  // Estado y respuestas
  final String status;
  final String? auditorId;
  final String? auditorResponse;
  final int? approvedDays;
  final bool needsAudit;
  final bool? isJustified;
  final DateTime? auditDate;
  
  // Archivos adjuntos
  final List<String>? attachments;
  
  // Metadata
  final String createdBy;
  final String? lastModifiedBy;
  final DateTime createdAt;
  final DateTime updatedAt;

  MedicalCertificate({
    required this.id,
    required this.userId,
    this.certificateNumber,
    required this.issueDate,
    required this.startDate,
    required this.endDate,
    required this.requestedDays,
    this.diagnosisCode,
    this.diagnosis,
    this.symptoms,
    required this.hasVisitedDoctor,
    this.medicalCenter,
    this.attendingPhysician,
    this.medicalPrescription,
    this.questionnaire,
    required this.status,
    this.auditorId,
    this.auditorResponse,
    this.approvedDays,
    required this.needsAudit,
    this.isJustified,
    this.auditDate,
    this.attachments,
    required this.createdBy,
    this.lastModifiedBy,
    required this.createdAt,
    required this.updatedAt,
  });

  factory MedicalCertificate.fromJson(Map<String, dynamic> json) {
    return MedicalCertificate(
      id: json['id'],
      userId: json['userId'],
      certificateNumber: json['certificateNumber'],
      issueDate: DateTime.parse(json['issueDate']),
      startDate: DateTime.parse(json['startDate']),
      endDate: DateTime.parse(json['endDate']),
      requestedDays: json['requestedDays'],
      diagnosisCode: json['diagnosisCode'],
      diagnosis: json['diagnosis'],
      symptoms: json['symptoms'],
      hasVisitedDoctor: json['hasVisitedDoctor'] ?? false,
      medicalCenter: json['medicalCenter'],
      attendingPhysician: json['attendingPhysician'],
      medicalPrescription: json['medicalPrescription'],
      questionnaire: json['questionnaire'] != null 
          ? Map<String, dynamic>.from(json['questionnaire'])
          : null,
      status: json['status'],
      auditorId: json['auditorId'],
      auditorResponse: json['auditorResponse'],
      approvedDays: json['approvedDays'],
      needsAudit: json['needsAudit'] ?? true,
      isJustified: json['isJustified'],
      auditDate: json['auditDate'] != null 
          ? DateTime.parse(json['auditDate'])
          : null,
      attachments: json['attachments'] != null 
          ? List<String>.from(json['attachments'])
          : null,
      createdBy: json['createdBy'],
      lastModifiedBy: json['lastModifiedBy'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'certificateNumber': certificateNumber,
      'issueDate': issueDate.toIso8601String(),
      'startDate': startDate.toIso8601String(),
      'endDate': endDate.toIso8601String(),
      'requestedDays': requestedDays,
      'diagnosisCode': diagnosisCode,
      'diagnosis': diagnosis,
      'symptoms': symptoms,
      'hasVisitedDoctor': hasVisitedDoctor,
      'medicalCenter': medicalCenter,
      'attendingPhysician': attendingPhysician,
      'medicalPrescription': medicalPrescription,
      'questionnaire': questionnaire,
      'status': status,
      'auditorId': auditorId,
      'auditorResponse': auditorResponse,
      'approvedDays': approvedDays,
      'needsAudit': needsAudit,
      'isJustified': isJustified,
      'auditDate': auditDate?.toIso8601String(),
      'attachments': attachments,
      'createdBy': createdBy,
      'lastModifiedBy': lastModifiedBy,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  // Helper methods
  String get statusText {
    switch (status) {
      case 'pending':
        return 'Pendiente de revisión';
      case 'under_review':
        return 'En revisión médica';
      case 'approved':
        return 'Aprobado';
      case 'rejected':
        return 'Rechazado';
      case 'needs_audit':
        return 'Requiere auditoría';
      default:
        return 'Estado desconocido';
    }
  }

  bool get isPending => status == 'pending';
  bool get isUnderReview => status == 'under_review';
  bool get isApproved => status == 'approved';
  bool get isRejected => status == 'rejected';
  bool get requiresAudit => status == 'needs_audit';

  int get daysRemaining {
    final now = DateTime.now();
    if (isApproved && endDate.isAfter(now)) {
      return endDate.difference(now).inDays + 1;
    }
    return 0;
  }
}